[Upgrade Assistant] Add support for API keys when reindexing

alisonelizabeth opened the pull request:
Fixes https://github.com/elastic/kibana/issues/72014

Upon receiving a request to reindex, Upgrade Assistant stores the requester's credentials in memory and uses those credentials when updating saved objects and issuing requests for other reindexing-related operations. In a token-based authentication mechanism, these tokens have an expiration time that could expire before reindexing work completes causing the task to stall. 

As a partial fix, this PR implements granting a temporary API key that is stored in memory that is then used for issuing requests as advised via https://github.com/elastic/kibana/issues/72014#issuecomment-892394832. Once the reindex operation is completed, the API key is invalidated. Note: this implementation is only possible if API keys are enabled. If API keys are *not* enabled, we will have to continue using the current approach.

**How to test**

Unzip [6.8.16-data.zip](https://github.com/elastic/kibana/files/7007813/6.8.16-data.zip). This contains indices created on 6.8 to trigger the reindex action in UA.

1. **Test reindexing with API keys enabled.** To enable API keys, start ES as follows:

      ```
      yarn es snapshot -E xpack.security.authc.api_key.enabled=true -E path.data=/path/to/6.8.16-data
      ```

     - Navigate to UA --> Elasticsearch deprecations
     - Select an index that needs to be reindexed, and start the reindexing process
     - Once started, you can verify an API key was created by navigating to Security --> API keys
     - Verify the reindex process completed successfully 
     - Once the reindex process is complete, the API key should no longer appear under Security --> API keys

2. **Test reindexing with API keys disabled.** Restart ES:

      ```
      yarn es snapshot -E path.data=/path/to/6.8.16-data
      ```

     - Navigate to UA --> Elasticsearch deprecations
     - Select an index that needs to be reindexed, and start the reindexing process
     - Verify the reindex process completed successfully

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/)
* Commit: 7f46f7ea6a84e29265a756b8455ac45f8ebe39fe
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types and Public API Docs](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/execution/node/428/log)

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/","number":"151617","commit":"7f46f7ea6a84e29265a756b8455ac45f8ebe39fe"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/)
* Commit: 68546805df23f4793f588c68db21191bcec229b9
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types and Public API Docs](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/execution/node/421/log)

### History
* :broken_heart: [Build #151617](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/) failed 7f46f7ea6a84e29265a756b8455ac45f8ebe39fe

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/","number":"151617","commit":"7f46f7ea6a84e29265a756b8455ac45f8ebe39fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/","number":"151900","commit":"68546805df23f4793f588c68db21191bcec229b9"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/)
* Commit: 8741beeca632a2d9086404fc59281b87fed59481
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types and Public API Docs](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/execution/node/428/log)

### History
* :broken_heart: [Build #151900](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/) failed 68546805df23f4793f588c68db21191bcec229b9
* :broken_heart: [Build #151617](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/) failed 7f46f7ea6a84e29265a756b8455ac45f8ebe39fe

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/","number":"151617","commit":"7f46f7ea6a84e29265a756b8455ac45f8ebe39fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/","number":"151900","commit":"68546805df23f4793f588c68db21191bcec229b9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/","number":"152034","commit":"8741beeca632a2d9086404fc59281b87fed59481"}]}
PIPELINE-->

kibanamachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/)
* Commit: db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde
* Storybooks not built yet
* This comment will update when the build is complete

### Failed CI Steps
* [Check Types and Public API Docs](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/execution/node/428/log)

### History
* :broken_heart: [Build #152034](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/) failed 8741beeca632a2d9086404fc59281b87fed59481
* :broken_heart: [Build #151900](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/) failed 68546805df23f4793f588c68db21191bcec229b9
* :broken_heart: [Build #151617](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/) failed 7f46f7ea6a84e29265a756b8455ac45f8ebe39fe

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/","number":"151617","commit":"7f46f7ea6a84e29265a756b8455ac45f8ebe39fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/","number":"151900","commit":"68546805df23f4793f588c68db21191bcec229b9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/","number":"152034","commit":"8741beeca632a2d9086404fc59281b87fed59481"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/","number":"152307","commit":"db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152329/)
* Commit: c03bbb9479a57bf768cf2aba1c100917c2117cc0
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111451/c03bbb9479a57bf768cf2aba1c100917c2117cc0)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :broken_heart: [Build #152307](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/) failed db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde
* :broken_heart: [Build #152034](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/) failed 8741beeca632a2d9086404fc59281b87fed59481
* :broken_heart: [Build #151900](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/) failed 68546805df23f4793f588c68db21191bcec229b9
* :broken_heart: [Build #151617](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/) failed 7f46f7ea6a84e29265a756b8455ac45f8ebe39fe

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/","number":"151617","commit":"7f46f7ea6a84e29265a756b8455ac45f8ebe39fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/","number":"151900","commit":"68546805df23f4793f588c68db21191bcec229b9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/","number":"152034","commit":"8741beeca632a2d9086404fc59281b87fed59481"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/","number":"152307","commit":"db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152329/","number":"152329","commit":"c03bbb9479a57bf768cf2aba1c100917c2117cc0"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152419/)
* Commit: 2d26060de0c1d74aa79a0f7131368f9155915717
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111451/2d26060de0c1d74aa79a0f7131368f9155915717)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #152329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152329/) succeeded c03bbb9479a57bf768cf2aba1c100917c2117cc0
* :broken_heart: [Build #152307](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/) failed db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde
* :broken_heart: [Build #152034](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/) failed 8741beeca632a2d9086404fc59281b87fed59481
* :broken_heart: [Build #151900](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/) failed 68546805df23f4793f588c68db21191bcec229b9
* :broken_heart: [Build #151617](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/) failed 7f46f7ea6a84e29265a756b8455ac45f8ebe39fe

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151617/","number":"151617","commit":"7f46f7ea6a84e29265a756b8455ac45f8ebe39fe"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/","number":"151900","commit":"68546805df23f4793f588c68db21191bcec229b9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/","number":"152034","commit":"8741beeca632a2d9086404fc59281b87fed59481"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/","number":"152307","commit":"db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152329/","number":"152329","commit":"c03bbb9479a57bf768cf2aba1c100917c2117cc0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152419/","number":"152419","commit":"2d26060de0c1d74aa79a0f7131368f9155915717"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/kibana-stack-management (Team:Stack Management)

jportner commented on x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts:
```diff
@@ -20,25 +23,119 @@ export type Credential = Record<string, any>;
+    security?: SecurityPluginStart;
+    reindexOpId: string;
+  }): Promise<string | undefined> => {
+    const apiKeyResult = await security?.authc.apiKeys.grantAsInternalUser(request, {
+      name: `ua_reindex_${reindexOpId}`,
+      role_descriptors: {},
+    });
```
Optional nit: given the chance that an API key could get orphaned, maybe it's worth adding a bit of metadata? Something like this:

```diff
diff --git a/x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts b/x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts
index 7f3f4f2b307..41eb7a5afa7 100644
--- a/x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts
+++ b/x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts
@@ -59,6 +59,10 @@ export const credentialStoreFactory = (logger: Logger): CredentialStore => {
     const apiKeyResult = await security?.authc.apiKeys.grantAsInternalUser(request, {
       name: `ua_reindex_${reindexOpId}`,
       role_descriptors: {},
+      metadata: {
+        description:
+          'Created by the Upgrade Assistant for a reindex operation; this can be safely deleted after Kibana is upgraded.',
+      },
     });
 
     if (apiKeyResult) {
```

alisonelizabeth replied:
Great point! Done.

jportner reviewed (commented):
I tested this as described, and also with the Security plugin disabled in ES.
Everything worked as expected! 🎉 

I also double checked to make sure we aren't passing _all_ client request headers to ES -- this plugin already uses the scoped cluster client provided by Core, which is the correct approach.

I have just a couple of nits below, great job on this.

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/server/lib/reindexing/credential_store.ts:
```diff
@@ -20,25 +23,119 @@ export type Credential = Record<string, any>;
+      // This can be used when sending a request with an "Authorization: ApiKey xxx" header
+      return Buffer.from(`${id}:${apiKey}`).toString('base64');
+    }
+  };
```
Thanks for the background! I've added a try/catch, which falls back to the user's credentials.

jportner reviewed (approved):
LGTM!

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152886/)
* Commit: 1b2bd0584efd179703cbbbf79b044206d4c83593
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111451/1b2bd0584efd179703cbbbf79b044206d4c83593)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #152419](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152419/) succeeded 2d26060de0c1d74aa79a0f7131368f9155915717
* :green_heart: [Build #152329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152329/) succeeded c03bbb9479a57bf768cf2aba1c100917c2117cc0
* :broken_heart: [Build #152307](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/) failed db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde
* :broken_heart: [Build #152034](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/) failed 8741beeca632a2d9086404fc59281b87fed59481
* :broken_heart: [Build #151900](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/) failed 68546805df23f4793f588c68db21191bcec229b9

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/151900/","number":"151900","commit":"68546805df23f4793f588c68db21191bcec229b9"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/","number":"152034","commit":"8741beeca632a2d9086404fc59281b87fed59481"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/","number":"152307","commit":"db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152329/","number":"152329","commit":"c03bbb9479a57bf768cf2aba1c100917c2117cc0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152419/","number":"152419","commit":"2d26060de0c1d74aa79a0f7131368f9155915717"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152886/","number":"152886","commit":"1b2bd0584efd179703cbbbf79b044206d4c83593"}]}
PIPELINE-->

cjcenizal reviewed (approved):
Tested locally, code LGTM! Thanks for adding tests. Had some nitpicky suggestions, nothing major.

@alisonelizabeth One thing I was missing was a way to reproduce the original problem raised by #72014, so I could verify that these changes address it. Do you know if there's a way to do that?

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/153063/)
* Commit: b005da7dd4dccb3d459fc750b2a16530a185e255
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111451/b005da7dd4dccb3d459fc750b2a16530a185e255)
* [Documentation Changes](https://kibana_111451.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #152886](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152886/) succeeded 1b2bd0584efd179703cbbbf79b044206d4c83593
* :green_heart: [Build #152419](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152419/) succeeded 2d26060de0c1d74aa79a0f7131368f9155915717
* :green_heart: [Build #152329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152329/) succeeded c03bbb9479a57bf768cf2aba1c100917c2117cc0
* :broken_heart: [Build #152307](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/) failed db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde
* :broken_heart: [Build #152034](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/) failed 8741beeca632a2d9086404fc59281b87fed59481

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152034/","number":"152034","commit":"8741beeca632a2d9086404fc59281b87fed59481"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/","number":"152307","commit":"db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152329/","number":"152329","commit":"c03bbb9479a57bf768cf2aba1c100917c2117cc0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152419/","number":"152419","commit":"2d26060de0c1d74aa79a0f7131368f9155915717"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152886/","number":"152886","commit":"1b2bd0584efd179703cbbbf79b044206d4c83593"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/153063/","number":"153063","commit":"b005da7dd4dccb3d459fc750b2a16530a185e255"}]}
PIPELINE-->

alisonelizabeth commented:
Thanks for the review @cjcenizal! I addressed your feedback.

> @alisonelizabeth One thing I was missing was a way to reproduce the original problem raised by #72014, so I could verify that these changes address it. Do you know if there's a way to do that?

It would require configuring SAML, which I believe is dependent on having an Identity Provider. I'm reaching out to the security team to see what options we might have.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/153190/)
* Commit: 01cca25539b14d0469a358315b98bc8a36e0b17e
* [Storybooks Preview](https://ci-artifacts.kibana.dev/storybooks/pr-111451/01cca25539b14d0469a358315b98bc8a36e0b17e)
* [Documentation Changes](https://kibana_111451.docs-preview.app.elstc.co/diff)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #153063](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/153063/) succeeded b005da7dd4dccb3d459fc750b2a16530a185e255
* :green_heart: [Build #152886](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152886/) succeeded 1b2bd0584efd179703cbbbf79b044206d4c83593
* :green_heart: [Build #152419](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152419/) succeeded 2d26060de0c1d74aa79a0f7131368f9155915717
* :green_heart: [Build #152329](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152329/) succeeded c03bbb9479a57bf768cf2aba1c100917c2117cc0
* :broken_heart: [Build #152307](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/) failed db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152307/","number":"152307","commit":"db9d5a6d0cc76497bf4c0e4b070cb5e3fdf5cbde"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152329/","number":"152329","commit":"c03bbb9479a57bf768cf2aba1c100917c2117cc0"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152419/","number":"152419","commit":"2d26060de0c1d74aa79a0f7131368f9155915717"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/152886/","number":"152886","commit":"1b2bd0584efd179703cbbbf79b044206d4c83593"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/153063/","number":"153063","commit":"b005da7dd4dccb3d459fc750b2a16530a185e255"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/153190/","number":"153190","commit":"01cca25539b14d0469a358315b98bc8a36e0b17e"}]}
PIPELINE-->

alisonelizabeth commented:
I opened https://github.com/elastic/kibana/pull/112154 to document testing instructions.

I reproduced the behavior on 7.x by setting a short token auth timeout (`-E xpack.security.authc.token.timeout=5s`) and verified reindexing is now working as expected with my changes.

<img width="499" alt="Screen Shot 2021-09-14 at 12 50 13 PM" src="https://user-images.githubusercontent.com/5226211/133307512-d2d0f9e7-7df7-42d3-b9e8-1ff802e09c30.png">
